trees/utils/StrongTreeFairOCT.py

```diff
@@ -0,0 +1,284 @@
+'''
+This module formulate the FlowOCT problem in gurobipy.
+'''
+
+from gurobipy import *
+import numpy as np
+import pandas as pd
+from itertools import combinations
+
+
+class FairOCT:
+    def __init__(self, X, y, tree, X_col_labels, labels, _lambda, time_limit, num_threads,
+                 fairness_type, fairness_bound,
+                 positive_class,
+                 P, P_col_labels, L, L_col_labels):
+        '''
+        :param X: numpy matrix or pandas data-frame of covariates.
+                  It's up to the user to include the protected features in X or not.
+                  We assume that we are allowed to branch on any of the columns within X.
+        :param y: numpy array or pandas series/data-frame of class labels
+        :param tree: Tree object
+        :param X_col_labels: The column names of matrix X
+        :param labels: the unique values of y
+        :param _lambda: The regularization parameter in the objective
+        :param time_limit: The given time limit for solving the MIP
+        :param num_threads: Number of threads for the solver to use
+        :param fairness_type: The type of the fairness constraint you wish to enforce, e.g., SP
+        :param fairness_bound: The fairness bound,  a real value between (0,1]
+        :param positive_class:
+        :param P: P Is the np.array of the protected features. Its dimension is (n_sample, n_p) where n_p is number of
+                  protected feaures.
+        :param L: L Is the np.array of the legitimate features. Its dimension is (n_sample, n_l) where n_l is number
+                  of legitimate features.
+        :param P_col_labels: Names of the protected columns
+        :param L_col_labels: Names of the legitimate columns
+        '''
+
+        self.X = pd.DataFrame(X, columns=X_col_labels)
+        self.y = y
+        self.P = P
+        self.L = L
+
+        self.X_p = np.concatenate((P, L, y.reshape(-1,1)), axis=1)
+        self.X_p = pd.DataFrame(self.X_p, columns= (P_col_labels.tolist() + L_col_labels.tolist() + ['y']))
+
+        self.X_col_labels = X_col_labels
+        self.labels = labels
+
+        self.P_col_labels = P_col_labels
+        self.L_col_labels = L_col_labels
+
+        # datapoints contains the indicies of our training data
+        self.datapoints = np.arange(0, self.X.shape[0])
+
+        self.tree = tree
+        self._lambda = _lambda
+
+        self.fairness_type = fairness_type
+        self.fairness_bound = fairness_bound
+        self.positive_class = positive_class
+
+
+        # Decision Variables
+        self.b = 0
+        self.p = 0
+        self.w = 0
+        self.zeta = 0
+        self.z = 0
+
+        # Gurobi model
+        self.model = Model('FairOCT')
+        self.model.params.Threads = num_threads
+        self.model.params.TimeLimit = time_limit
+
+
+
+
+    def create_primal_problem(self):
```

Comment:
For some formulations we have a `create_primal_problem` and for others we have `create_master_problem`. Can we standardize them to all be `create_master_problem`?